core/src/main/java/ai/timefold/solver/core/impl/bavet/common/AbstractStaticDataNode.java

```diff
@@ -0,0 +1,159 @@
+package ai.timefold.solver.core.impl.bavet.common;
+
+import java.util.ArrayList;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Map;
+
+import ai.timefold.solver.core.impl.bavet.NodeNetwork;
+import ai.timefold.solver.core.impl.bavet.common.tuple.AbstractTuple;
+import ai.timefold.solver.core.impl.bavet.common.tuple.RecordingTupleLifecycle;
+import ai.timefold.solver.core.impl.bavet.common.tuple.TupleLifecycle;
+import ai.timefold.solver.core.impl.bavet.common.tuple.TupleState;
+import ai.timefold.solver.core.impl.bavet.uni.AbstractForEachUniNode;
+
+import org.jspecify.annotations.NullMarked;
+import org.jspecify.annotations.Nullable;
+
+@NullMarked
+public abstract class AbstractStaticDataNode<Tuple_ extends AbstractTuple> extends AbstractNode
+        implements TupleSourceRoot<Object> {
+    private final StaticPropagationQueue<Tuple_> propagationQueue;
+    private final Map<Object, List<Tuple_>> tupleMap = new IdentityHashMap<>(1000);
+    private final NodeNetwork nodeNetwork;
+    private final RecordingTupleLifecycle<Tuple_> recordingTupleNode;
+    private final Class<?>[] sourceClasses;
+
+    public AbstractStaticDataNode(NodeNetwork nodeNetwork,
+            RecordingTupleLifecycle<Tuple_> recordingTupleNode,
+            TupleLifecycle<Tuple_> nextNodesTupleLifecycle,
+            Class<?>[] sourceClasses) {
+        this.nodeNetwork = nodeNetwork;
+        this.propagationQueue = new StaticPropagationQueue<>(nextNodesTupleLifecycle);
+        this.sourceClasses = sourceClasses;
+        this.recordingTupleNode = recordingTupleNode;
+    }
+
+    @Override
+    public final Propagator getPropagator() {
+        return propagationQueue;
+    }
+
+    @Override
+    public final boolean allowsInstancesOf(Class<?> clazz) {
+        for (var sourceClass : sourceClasses) {
+            if (sourceClass.isAssignableFrom(clazz)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public final Class<?>[] getSourceClasses() {
+        return sourceClasses;
+    }
+
+    @Override
+    public final boolean supports(TupleSourceRoot.LifecycleOperation lifecycleOperation) {
+        return true;
+    }
+
+    @Override
+    public final void insert(Object a) {
+        if (tupleMap.containsKey(a)) {
+            return;
+        }
+        invalidateCache();
+        tupleMap.put(a, new ArrayList<>());
+        insertIntoNodeNetwork(a);
+        recalculateTuples();
```

Comment:
Alternative implementation suggestion:

Inserts/updates/retracts go to a map.
If an update lands and an insert is already there, the update is ignored.
If an update lands twice, we only keep the first one.
if a retract lands and an insert is already there, the insert is removed and the retract is ignored.

All the operations are then pushed all at once, before propagation starts.
This way, we minimize the amount of operations that need to be performed, and we also group them in time, nothing happens before propagation; which would likely help with sizing collections etc.